src/test/java/com/google/sps/managers/DatastoreManagerTest.java

```diff
@@ -0,0 +1,301 @@
+  }
+
+  @Test
+  public void createOrReplaceStageAndRetrieveTest() {
+    Stage expected;
+    Stage actual;
+    
+    //Makes use of the Stage class and it's functions defined previously to build a single stage for testing purposes
+    Stage.Builder stage = new Stage.Builder("0001", 4);
+    stage.setKey("Test").setStartingHint("Testing").setStartingLocation(new Coordinates()).setHints(new ArrayList<Hint>());
+    expected = stage.build();
+
+    datastoreManager.createOrReplaceStage(expected);
+
+    actual = datastoreManager.retrieveStage("0001");
+
+    Assert.assertEquals(expected.getStageID(), actual.getStageID());
+  }
+
+  @Test
+  public void createOrReplaceSinglePlayerProgressAndRetrieveTest() {
+    SinglePlayerProgress expected;
+    SinglePlayerProgress actual;
+    
+    //Makes use of the SinglePlayerProgress class defined previously to build a single SinglePlayerProgress for testing purposes
+    SinglePlayerProgress.Builder progress = new SinglePlayerProgress.Builder("0000", "0001");
+    progress.setLocation(new Coordinates()).setHintsFound(new ArrayList<Integer>()).setStageID("0002");
+    expected = progress.build();
+
+    datastoreManager.createOrReplaceSinglePlayerProgress(expected);
+
+    actual = datastoreManager.retrieveSinglePlayerProgress("0000", "0001");
+
+    Assert.assertEquals(expected.getUserID(), actual.getUserID());
+  }
+
+  @Test
+  public void createOrReplaceHintAndRetrieveTest() {
+    Hint expected;
+    Hint actual;
+    
+    //Makes use of the Hint class defined previously to build a single Hint for testing purposes
+    Hint.Builder hint = new Hint.Builder("0001", 3);
+    hint.setLocation(new Coordinates()).setText("For testing");
+    expected = hint.build();
+
+    datastoreManager.createOrReplaceHint(expected);
+
+    actual = datastoreManager.retrieveHint("0001");
+
+    Assert.assertEquals(expected.getHintID(), actual.getHintID());
+  }
+
+  @Test
+  public void retrieveAllGamesTest() {
+    int expected = 10;
+    int actual;
+    ArrayList<Game> test = new ArrayList<>();
+  
+    for(int i = 0; i < 10; i++) test.add(getRandomGame());
+    actual =  test.size();
+
+    Assert.assertEquals(expected, actual);
+  }
+  
+  @Test
+  public void doesUsernameExistTest() {
+    User test;
+    boolean actual;
+    boolean expected = true;
+    
+    //Makes use of the User class defined previously to build a single user for testing purposes
+    User.Builder user = new User.Builder("0001").setUsername("user1").setFirstName("Ade").setLastName("Ademiluyi");
+    user.setProfilePictureUrl("images/defaultProfilePicture.jpg").setGamesCreated(new ArrayList<String>());
+    user.setGamesCompletedWithTime(new ArrayList<Pair<String, Long>>());
+    test = user.build();
+
+    datastoreManager.createOrReplaceUser(test);
+
+    actual = datastoreManager.doesUsernameExist("user1");
+
+    Assert.assertEquals(expected, actual);
+  }
+
+  @Test
+  public void deleteSinglePlayerProgressTest() {
```

Comment:
Add a case where you are trying to delete a SinglePlayerProgress that doesn't exist. In this case you should probably throw an error.

Reply:
sure. already wroking on testing edge cases